File: rateLimiter/fastapi-service/app/routers/action.py
from typing import Optional

from fastapi import APIRouter, Header
from fastapi.responses import JSONResponse

from app.limiter.base import RateLimitResult
from app.registry import LIMITERS

router = APIRouter(prefix="/api")


def _rate_limit_headers(result: RateLimitResult) -> dict:
    headers = {
        "X-RateLimit-Limit": str(result.limit),
        "X-RateLimit-Remaining": str(max(0, result.remaining)),
        "X-RateLimit-Reset": str(result.reset_at),
    }
    if result.retry_after is not None:
        headers["Retry-After"] = str(result.retry_after)
    return headers


async def _handle(algorithm: str, x_client_id: Optional[str]) -> JSONResponse:
    if not x_client_id:
        return JSONResponse(status_code=400, content={"error": "missing_client_id"})

    result = await LIMITERS[algorithm].allow(x_client_id)
    headers = _rate_limit_headers(result)

    if not result.allowed:
        return JSONResponse(status_code=429, content={"error": "rate_limited"}, headers=headers)

    return JSONResponse(status_code=200, content={"ok": True, "algorithm": algorithm}, headers=headers)


@router.post("/token-bucket/action")
async def token_bucket_action(x_client_id: Optional[str] = Header(default=None, alias="X-Client-Id")):
    return await _handle("token_bucket", x_client_id)


@router.post("/leaky-bucket/action")
async def leaky_bucket_action(x_client_id: Optional[str] = Header(default=None, alias="X-Client-Id")):
    return await _handle("leaky_bucket", x_client_id)


@router.post("/fixed-window/action")
async def fixed_window_action(x_client_id: Optional[str] = Header(default=None, alias="X-Client-Id")):
    return await _handle("fixed_window", x_client_id)


@router.post("/sliding-window-log/action")
async def sliding_window_log_action(x_client_id: Optional[str] = Header(default=None, alias="X-Client-Id")):
    return await _handle("sliding_window_log", x_client_id)


@router.post("/sliding-window-counter/action")
async def sliding_window_counter_action(
    x_client_id: Optional[str] = Header(default=None, alias="X-Client-Id")
):
    return await _handle("sliding_window_counter", x_client_id)

File: rateLimiter/go-service/internal/handlers/action.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"

	"ratelimiter/internal/limiter"
)

type ActionHandler struct {
	Limiters map[string]limiter.Limiter
}

func writeJSON(w http.ResponseWriter, status int, v interface{}, headers map[string]string) {
	for k, val := range headers {
		w.Header().Set(k, val)
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func rateLimitHeaders(r limiter.Result) map[string]string {
	h := map[string]string{
		"X-RateLimit-Limit":     fmt.Sprintf("%d", r.Limit),
		"X-RateLimit-Remaining": fmt.Sprintf("%d", max(0, r.Remaining)),
		"X-RateLimit-Reset":     fmt.Sprintf("%v", r.ResetAt),
	}
	if r.RetryAfter != nil {
		h["Retry-After"] = fmt.Sprintf("%v", *r.RetryAfter)
	}
	return h
}

func (h *ActionHandler) Handle(algorithm string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		clientID := r.Header.Get("X-Client-Id")
		if clientID == "" {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "missing_client_id"}, nil)
			return
		}

		result, err := h.Limiters[algorithm].Allow(r.Context(), clientID)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "internal_error"}, nil)
			return
		}

		headers := rateLimitHeaders(result)
		if !result.Allowed {
			writeJSON(w, http.StatusTooManyRequests, map[string]string{"error": "rate_limited"}, headers)
			return
		}
		writeJSON(w, http.StatusOK, map[string]interface{}{"ok": true, "algorithm": algorithm}, headers)
	}
}
